rust/worker/src/segment/types.rs

```diff
@@ -1,29 +1,362 @@
+    MetadataValueConversionError, Operation, OperationRecord,
+};
 use async_trait::async_trait;
+use thiserror::Error;
 
-#[derive(Debug)]
+use super::record_segment::{ApplyMaterializedLogError, RecordSegmentReader};
+
+#[derive(Error, Debug)]
+pub enum LogMaterializerError {
+    #[error("Error materializing document metadata {0}")]
+    MetadataMaterializationError(#[from] MetadataValueConversionError),
+    #[error("Error materializing document embedding")]
+    EmbeddingMaterializationError,
+    #[error("Error reading record segment {0}")]
+    RecordSegmentError(#[from] Box<dyn ChromaError>),
+}
+
+impl ChromaError for LogMaterializerError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            LogMaterializerError::MetadataMaterializationError(e) => e.code(),
+            LogMaterializerError::EmbeddingMaterializationError => ErrorCodes::Internal,
+            LogMaterializerError::RecordSegmentError(e) => e.code(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
 pub(crate) struct MaterializedLogRecord<'a> {
-    pub(super) segment_offset_id: u32,
-    pub(super) log_record: &'a LogRecord,
-    pub(super) materialized_record: DataRecord<'a>,
+    // This is the data record read from the record segment for this id.
+    // None if the record exists only in the log.
+    pub(super) data_record: Option<DataRecord<'a>>,
+    // If present in the record segment then it is the offset id
+    // in the record segment at which the record was found.
+    // If not present in the segment then it is the offset id
+    // at which it should be inserted.
+    pub(super) offset_id: u32,
+    // Set only for the records that are being inserted for the first time
+    // in the log since data_record will be None in such cases. For other
+    // cases, just read from data record.
+    pub(super) user_id: Option<&'a str>,
+    // There can be several entries in the log for an id. This is the final
+    // operation that needs to be done on it. For e.g.
+    // If log has [Update, Update, Delete] then final operation is Delete.
+    // If log has [Insert, Update, Update, Delete] then final operation is Delete.
+    // If log has [Insert, Update, Update] then final operation is Insert.
+    // If log has [Update, Update] then final operation is Update.
+    // All Upserts are converted to Update or Insert as the final operation.
+    // For e.g. if log has [Insert, Upsert] then final operation is update.
+    // If log has [Upsert] and the record does not exist in storage then final
+    // operation is Insert.
+    pub(super) final_operation: Operation,
+    // This is the metadata obtained by combining all the operations
+    // present in the log for this id.
+    // E.g. if has log has [Insert(a: h), Update(a: b, c: d), Update(a: e, f: g)] then this
+    // will contain (a: e, c: d, f: g). This is None if the final operation
+    // above is Delete.
+    pub(super) metadata_to_be_merged: Option<Metadata>,
```

Comment:
nit: this name/structure is confusing to me, why do we not merge the metadata and also present it here?

Reply:
Discussed and resolved offline. This is fine